File: src/test/java/com/hp/jipp/encoding/NameCodeTest.java
package com.hp.jipp.encoding;

import org.junit.Test;

import static org.junit.Assert.*;

import static com.hp.jipp.encoding.Cycler.*;

import com.google.common.collect.ImmutableList;

public class NameCodeTest {
    /** An enumeration of possible printer states */
    public static class Sample extends NameCode {

        public static final Sample One = new Sample("one", 1);
        public static final Sample Two = new Sample("two", 2);
        public static final Sample Three = new Sample("three", 3);
        // Cannot be reached
        private static final Sample Secret = new Sample("secret", 4);

        public static final NameCodeType.Encoder<Sample> ENCODER = NameCodeType.Encoder.of(
                Sample.class, new NameCode.Factory<Sample>() {
                    @Override
                    public Sample of(String name, int code) {
                        return new Sample(name, code);
                    }
                });
        private final int code;
        private final String name;

        @Override
        public String getName() {
            return name;
        }

        @Override
        public int getCode() {
            return code;
        }

        Sample(String name, int code) {
            this.name = name;
            this.code = code;
        }
    }
    NameCodeType<Sample> MySample = NameCodeType.typeOf(Sample.ENCODER, "my-sample");

    @Test
    public void sample() throws Exception {
        assertEquals(ImmutableList.of(Sample.One), cycle(MySample, MySample.of(Sample.One)).getValues());
    }

    @Test
    public void fetchFromGroup() throws Exception {
        assertEquals(ImmutableList.of(Sample.Two, Sample.Three),
                cycle(AttributeGroup.of(Tag.JobAttributes, MySample.of(Sample.Two, Sample.Three))).getValues(MySample));
    }
}

File: src/test/java/com/hp/jipp/encoding/NameCodeTest.kt
package com.hp.jipp.encoding

import org.junit.Test

import org.junit.Assert.*

import com.hp.jipp.encoding.Cycler.*

import com.google.common.collect.ImmutableList
import com.hp.jipp.util.KotlinTest
import org.hamcrest.CoreMatchers.*

class NameCodeTest {
    /** An enumeration of possible printer states  */
    class Sample internal constructor(override val name: String, override val code: Int) : NameCode() {
        companion object {

            // TODO: How to reflect on these?
            @JvmField val One = Sample("one", 1)
            @JvmField val Two = Sample("two", 2)
            @JvmField val Three = Sample("three", 3)

            // Cannot be reached
            private val Secret = Sample("secret", 4)

            val ENCODER: NameCodeType.Encoder<Sample> = NameCodeType.Encoder.of(
                    Sample::class.java, object : NameCode.Factory<Sample> {
                override fun of(name: String, code: Int): Sample {
                    return Sample(name, code)
                }
            })
        }
    }

    internal var MySample = NameCodeType(Sample.ENCODER, "my-sample")

    @Test
    @Throws(Exception::class)
    fun sample() {
        assertEquals(ImmutableList.of(Sample.One), cycle(MySample, MySample.of(Sample.One)).values)
        assertThat(NameCode.allFrom<Sample>(Sample::class.java), hasItems(Sample.One, Sample.Two, Sample.Three))
    }

    @Test
    @Throws(Exception::class)
    fun fetchFromGroup() {
        assertEquals(ImmutableList.of(Sample.Two, Sample.Three),
                cycle(AttributeGroup.of(Tag.JobAttributes, MySample.of(Sample.Two, Sample.Three))).getValues(MySample))
    }

    @Test
    @Throws(Exception::class)
    fun cover() {
        val encoder = MySample.nameCodeEncoder
        KotlinTest.cover(encoder,
                encoder.copy(encoder.type, encoder.map, encoder.factory),
                encoder.copy("other", encoder.map, encoder.factory))
    }
}
